src/range_proof.rs

```diff
@@ -0,0 +1,558 @@
+        self.r1
+    }
+
+    fn get_s1(&self) -> Scalar {
+        self.s1
+    }
+
+    fn get_d1(&self) -> Scalar {
+        self.d1
+    }
+
+    fn get_li(&self) -> Result<Vec<RistrettoPoint>, ProofError> {
+        if !self.li.is_empty() {
+            let mut li = Vec::with_capacity(self.li.len());
+            for item in self.li.clone() {
+                li.push(item.decompress().ok_or_else(|| {
+                    ProofError::InternalDataInconsistent(
+                        "An item in member 'L' was not the canonical encoding of a point".to_string(),
+                    )
+                })?)
+            }
+            Ok(li)
+        } else {
+            Err(ProofError::InternalDataInconsistent(
+                "Vector 'L' not assigned yet".to_string(),
+            ))
+        }
+    }
+
+    fn get_ri(&self) -> Result<Vec<RistrettoPoint>, ProofError> {
+        if !self.li.is_empty() {
+            let mut ri = Vec::with_capacity(self.ri.len());
+            for item in self.ri.clone() {
+                ri.push(item.decompress().ok_or_else(|| {
+                    ProofError::InternalDataInconsistent(
+                        "An item in member 'R' was not the canonical encoding of a point".to_string(),
+                    )
+                })?)
+            }
+            Ok(ri)
+        } else {
+            Err(ProofError::InternalDataInconsistent(
+                "Vector 'R' not assigned yet".to_string(),
+            ))
+        }
+    }
+
+    /// Create a single or aggregated range proof for a single party that knows all the secrets
+    pub fn prove(
+        transcript: &mut Transcript,
+        statement: &RangeStatement,
+        witness: RangeWitness,
+    ) -> Result<RangeProof, ProofError> {
+        let batch_size = statement.commitments.len();
+        if witness.openings.len() != batch_size {
+            return Err(ProofError::InvalidLength(
+                "Invalid range statement - commitments and openings do not match!".to_string(),
+            ));
+        }
+        for j in 0..batch_size {
+            if statement.commitments[j] !=
+                statement
+                    .generators
+                    .pc_gens()
+                    .commit(Scalar::from(witness.openings[j].v), witness.openings[j].r)
+            {
+                return Err(ProofError::InternalDataInconsistent(
+                    "Invalid range statement - commitment and opening data do not match".to_string(),
+                ));
+            }
+        }
+
+        let bit_length = statement.generators.bit_length();
+
+        // Global generators
+        let h_base = statement.generators.h_base();
+        let g_base = statement.generators.g_base();
+        let hi_base = statement.generators.hi_base();
+        let gi_base = statement.generators.gi_base();
+
+        transcript.domain_separator(b"Bulletproof+", b"Range Proof");
+        transcript.validate_and_append_point(b"H", &h_base.compress())?;
```

Comment:
You can compress these once in the generators, and return the compressed data directly